corehq/util/couch_doc_processor.py

```diff
@@ -367,3 +380,60 @@ def _processing_complete(self):
         ))
         if self.skipped:
             print(DOCS_SKIPPED_WARNING.format(self.skipped))
+
+
+class BulkDocProcessor(CouchDocumentProcessor):
+    """Process couch docs for a single doc type in batches
+
+    The bulk doc processor will send a batch of documents to the document
+    processor. If the processor does not respond with True then
+    the iteration is halted. Restarting the iteration will start by
+    re-sending the previous chunk to the processor.
+
+    The size of the batches passed to the document processor will be
+    ``chunk_size - may vary depending on how they are being filtered by the
+    document processor but will never exceed ``chunk_size``.
+
+    :param doc_type_map: Dict containing a single `doc_type_name: model_class` pair.
+    :param doc_processor: A `BaseDocProcessor` object used to
+    process documents.
+    :param reset: Reset existing processor state (if any), causing all
+    documents to be reconsidered for processing, if this is true.
+    :param chunk_size: Maximum number of records to read from couch at
+    one time. It may be necessary to use a smaller chunk size if the
+    records being processed are very large and the default chunk size of
+    100 would exceed available memory.
+    """
+    def __init__(self, doc_type_map, doc_processor, reset=False, chunk_size=100):
+        assert len(doc_type_map) == 1
+        super(BulkDocProcessor, self).__init__(doc_type_map, doc_processor, reset=reset, chunk_size=chunk_size)
+        self.changes = []
+
+    def _process_doc(self, doc):
+        if self.doc_processor.should_process(doc):
+            self.changes.append(doc)
+
+        if self.visited % self.chunk_size == 0:
```

Comment:
Looks like this is trying to stay in lock-step with the chunks generated by the underlying `paginate_view` iterator. The connection seems fragile. The requirement that chunks generated here stay in sync with `paginate_view` chunks could make it hard to change the code between here and there without accidentally breaking that contract.